lib/awful/layout/dynamic/suit/manual.lua

```diff
@@ -0,0 +1,169 @@
+--
+-- @property reflow
+-- @param boolean
+
+--TODO insert_at_end (boolean)
+
+-- Find the lowest priority client wrapper and remove it from the layout.
+local function extract_wrapper(self, below_priority)
+    local loser_priority, loser = below_priority
+
+    for w, _, _, parent in self:by_client() do
+        if parent and parent.priority and parent.priority < loser_priority then
+            loser_priority, loser = parent.priority, w
+        end
+    end
+
+    if not loser then return end
+
+    self._private._remove_widgets(self, loser, true)
+
+    return loser
+end
+
+local function max_elements(w, self)
+    return type(w.max_elements) == "function"
+        and w.max_elements(self) or w.max_elements
+end
+
+local function priority(w, self)
+    return type(w.priority) == "function"
+        and w.priority(self) or w.priority
+end
+
+local function select_target(self)
+    local p_best, r_best, p_best_w, r_best_w = -math.huge, 0
+
+    for _, w in ipairs(self:get_all_children()) do
+        if w.add then
+            local avail_space = (not w.max_elements)
+                and math.huge or (max_elements(w, self) - #w:get_children())
+
+            local has_priority = w.priority ~= nil
+                and avail_space > 0 and p_best < priority(w, self)
+
+            if has_priority then
+                p_best, p_best_w = priority(w, self), w
+            elseif avail_space > r_best then
+                r_best, r_best_w = avail_space, w
+            end
+        end
+    end
+
+    return p_best_w or r_best_w or self
+end
+
+local function ctr(args)
+    local function create()
+        local main_layout = wibox.widget(args)
+
+        main_layout._private._remove_widgets = main_layout.remove_widgets
+        main_layout._private._add = main_layout.add
+
+        function main_layout:add(w, ...)
+            if not w then return end
+
+            local t = select_target(self)
+            local f = t == self and main_layout._private._add or t.add
+            f(t, w)
+            main_layout:add(...)
+        end
+
+        -- Check if something has to be done afterward
+        function main_layout:remove_widgets(w, ...)
+            if not w then return end
+            local other = {...}
+            local _, parent, _ = self:index(w, true)
+
+            -- Call the real function
+            main_layout._private._remove_widgets(self, w, other[#other] == true)
+
+            if parent.reflow then
+                local replacement = extract_wrapper(self, parent.priority or 0)
+                if replacement then
+                    local f = parent == self and self._private._add or parent.add
+                    f(parent, replacement)
+                end
+                --TODO
+            end
+
+            self:remove_widgets(...)
+        end
+
+        return main_layout
+    end
+
+    -- Pick a name for this layout
+    local name = type(args.name) == "function" and name(args) or name or "manual"
```

Comment:
This casts a Travis `undefined` warning. I would suggest:

```diff
-local name = type(args.name) == "function" and name(args) or name or "manual"
+local ntype = type(args.name)
+local name = (ntype == "function" and args.name()) or (ntype == "string" and name) or "manual"
```

to also assert that the name can only be a function or a string.